internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go

```diff
@@ -0,0 +1,148 @@
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1/pbproxystate"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func (b *Builder) BuildDestinations(destinations []*intermediate.Destination) *Builder {
+	for _, destination := range destinations {
+		if destination.Explicit != nil {
+			b.buildExplicitDestination(destination)
+		}
+	}
+
+	return b
+}
+
+func (b *Builder) buildExplicitDestination(destination *intermediate.Destination) *Builder {
+	clusterName := DestinationClusterName(destination.Explicit.DestinationRef, destination.Explicit.Datacenter, b.trustDomain)
+	statPrefix := DestinationStatPrefix(destination.Explicit.DestinationRef, destination.Explicit.Datacenter)
+
+	// All endpoints should have the same protocol as the endpoints controller ensures that is the case,
+	// so it's sufficient to read just the first endpoint.
+	if len(destination.ServiceEndpoints.Endpoints.Endpoints) > 0 {
+		// Get destination port so that we can configure this destination correctly based on its protocol.
+		destPort := destination.ServiceEndpoints.Endpoints.Endpoints[0].Ports[destination.Explicit.DestinationPort]
+
+		// Find the destination proxy's port.
+		// Endpoints refs will need to route to mesh port instead of the destination port as that
+		// is the port of the destination's proxy.
+		meshPortName := findMeshPort(destination.ServiceEndpoints.Endpoints.Endpoints[0].Ports)
+
+		if destPort != nil {
+			return b.addOutboundDestinationListener(destination.Explicit).
+				addRouter(clusterName, statPrefix, destPort.Protocol).
+				addCluster(clusterName, destination.Identities).
+				addEndpointsRef(clusterName, destination.ServiceEndpoints.Resource.Id, meshPortName)
+		}
+	}
+
+	return b
+}
+
+func (b *Builder) addOutboundDestinationListener(explicit *pbmesh.Upstream) *Builder {
+	listener := &pbproxystate.Listener{
+		Direction: pbproxystate.Direction_DIRECTION_OUTBOUND,
+	}
+
+	// Create outbound listener address.
+	switch explicit.ListenAddr.(type) {
+	case *pbmesh.Upstream_IpPort:
+		destinationAddr := explicit.ListenAddr.(*pbmesh.Upstream_IpPort)
+		listener.BindAddress = &pbproxystate.Listener_HostPort{
+			HostPort: &pbproxystate.HostPortAddress{
+				Host: destinationAddr.IpPort.Ip,
+				Port: destinationAddr.IpPort.Port,
+			},
+		}
+		listener.Name = DestinationListenerName(explicit.DestinationRef.Name, explicit.DestinationPort, destinationAddr.IpPort.Ip, destinationAddr.IpPort.Port)
+	case *pbmesh.Upstream_Unix:
+		destinationAddr := explicit.ListenAddr.(*pbmesh.Upstream_Unix)
+		listener.BindAddress = &pbproxystate.Listener_UnixSocket{
+			UnixSocket: &pbproxystate.UnixSocketAddress{
+				Path: destinationAddr.Unix.Path,
+				Mode: destinationAddr.Unix.Mode,
+			},
+		}
+		listener.Name = DestinationListenerName(explicit.DestinationRef.Name, explicit.DestinationPort, destinationAddr.Unix.Path, 0)
+	}
```

Comment:
should the `default` switch case log an error and not add a listener?

Reply:
Thinking a bit more about this, I think the error case should be handled by proxy state template validation on save rather than logging so that it can be better communicated to the user. In this case, if the bind type is not one of these options, we'll end up with a listener without a bind address which should be invalid.